zerver/lib/export.py

```diff
@@ -111,80 +386,42 @@ def export_with_admin_auth(realm, response, include_invite_only=True, include_pr
-        # some of these cross-realm messages.
-        huddle_subs = [sub for sub in realm_huddle_subs if sub.recipient_id not in unsafe_huddle_recipient_ids]
-        huddle_recipient_ids = set(sub.recipient_id for sub in huddle_subs)
-        huddle_ids = set(sub.recipient.type_id for sub in huddle_subs)
-
-        huddle_subscription_dicts = make_raw(huddle_subs)
-        huddle_recipients = make_raw(Recipient.objects.filter(id__in=huddle_recipient_ids))
-        response['zerver_huddle'] = make_raw(Huddle.objects.filter(id__in=huddle_ids))
-    else:
-        huddle_recipients = []
-        huddle_subscription_dicts = []
 
-    response["zerver_recipient"] = user_recipients + stream_recipients + huddle_recipients
-    response["zerver_subscription"] = user_subscription_dicts + stream_subscription_dicts + huddle_subscription_dicts
+def fetch_huddle_objects(response, config, context):
+    # type: (TableData, Config, Context) -> None
+    # First we get all huddles involving someone in the realm.
+    parent = config.parent
+    realm = context['realm']
+
+    user_profile_ids = set(r['id'] for r in response[parent.table])
+
+    realm_huddle_subs = Subscription.objects.select_related("recipient").filter(recipient__type=Recipient.HUDDLE,
+                                                                                user_profile__in=user_profile_ids)
+    realm_huddle_recipient_ids = set(sub.recipient_id for sub in realm_huddle_subs)
+
+    # Mark all Huddles whose recipient ID contains a cross-realm user.
+    unsafe_huddle_recipient_ids = set()
+    for sub in Subscription.objects.select_related().filter(recipient__in=realm_huddle_recipient_ids):
+        if sub.user_profile.realm != realm:
+            # In almost every case the other realm will be zulip.com
+            unsafe_huddle_recipient_ids.add(sub.recipient_id)
 
-    attachment_query = Attachment.objects.filter(realm=realm)
-    response["zerver_attachment"] = make_raw(attachment_query)
-    floatify_datetime_fields(response, 'zerver_attachment')
+    # Now filter down to just those huddles that are entirely within the realm.
+    #
+    # This is important for ensuring that the User objects needed
+    # to import it on the other end exist (since we're only
+    # exporting the users from this realm), at the cost of losing
+    # some of these cross-realm messages.
+    huddle_subs = [sub for sub in realm_huddle_subs if sub.recipient_id not in unsafe_huddle_recipient_ids]
+    huddle_recipient_ids = set(sub.recipient_id for sub in huddle_subs)
+    huddle_ids = set(sub.recipient.type_id for sub in huddle_subs)
+
+    huddle_subscription_dicts = make_raw(huddle_subs)
+    huddle_recipients = make_raw(Recipient.objects.filter(id__in=huddle_recipient_ids))
+
+    response['_huddle_recipient'] = huddle_recipients
+    response['_huddle_subscription'] = huddle_subscription_dicts
+    response['zerver_huddle'] = make_raw(Huddle.objects.filter(id__in=huddle_ids))
```

Comment:
can you split the refactoring of `fetch_huddle_objects` into a function into its own commit early in the branch?  I think that'll make this transition a lot more readable.